### PR TITLE
Recalculate model memory size

### PR DESCRIPTION
**Pull request by kaituo**
Note: since there are inter-file references, I split a big PR into multiple smaller PRs based on files to save time (1.1 cut off is due this Friday). The build will fail due to missing dependencies. I will use that PR just for review. will not merge it. Will have a big PR in the end and merge once after all review PRs get approved. For the complete code, check https://github.com/kaituo/anomaly-detection-1/tree/thresholdRCF2

### Description
This PR recalculates model memory size formula after introducing ThresholdedRandomCutForest.  Please check comments for details.

Testing done:
1. unit tests.

Experiment data:
https://gist.github.com/kaituo/77b985921eda631a534157e8e11f6c57
 
### Check List
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/ad/MemoryTracker.java [hunk 1]
```diff
@@ -238,42 +235,66 @@ public long estimateTotalModelSize(AnomalyDetector detector, int numberOfTrees,
...
-     *        - BoundingBoxFloat: (1040 + 255* (dimension * 4 * 2 + 64)) * adjusted bounding box cache usage,
-     *           where if full we have 255 inner node and each box has 80 bytes.
-     *           Plus metadata, we can have in total 21544 bytes.
-     *           {@code adjusted bounding box cache usage = (bounding box cache fraction >= 0.3? 1: bounding box cache fraction)}
-     *           {@code >= 0.3} we will still initialize bounding box cache array of the max size,
-     *           but exclude them using the cache ratio.  It is not guaranteed we will only
-     *           use cache ratio in the array.  For example, with cache ratio 0.5, we used 150
-     *           out of 255 elements.  So can have two float array whose size is the number of
-     *           dimensions; other constants are the metadata size.
+     *        - BoundingBoxFloat: (1040 + 255* (dimension * 4 + 16) * 2 + 32)) * actual bounding box cache usage,
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/ad/MemoryTracker.java:238]: Where is 1040 coming from?

File: src/main/java/org/opensearch/ad/MemoryTracker.java [hunk 2]
```diff
@@ -238,42 +235,66 @@ public long estimateTotalModelSize(AnomalyDetector detector, int numberOfTrees,
...
+     *           {@code actual bounding box cache usage = (bounding box cache fraction >= 0.3? 1: bounding box cache fraction)}
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/ad/MemoryTracker.java:239]: Why is there a ? after 0.3?

File: src/main/java/org/opensearch/ad/MemoryTracker.java [hunk 3]
```diff
@@ -238,42 +235,66 @@ public long estimateTotalModelSize(AnomalyDetector detector, int numberOfTrees,
...
+     *           {@code >= 0.3} we will still initialize bounding box cache array of the max size.
      * In total, RCF size is
-     *  56 + # trees * (2248 + 152 + 6120 + 104 + (1040 + 255* (dimension * 4 * 2 + 64)) * adjusted bounding box cache ratio) +
-     *  (256 * # trees  * 2 + 256 * # trees * dimension) * 4 bytes  * 0.5 + 1064 + 24 + 24 + 16
-     *  = 56 + # trees * (8624 + (1040 + 255 * (dimension * 8 + 64)) * adjusted bounding box cache ratio) + 256 * # trees *
-     *   (3 + dimension) * 4 * 0.5 + 1128
+     *  56 + # trees * (2248 + 152 + 6120 + 104 + (1040 + 255* (dimension * 4 + 16) * 2 + 32)) * adjusted bounding box cache ratio) +
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/ad/MemoryTracker.java:242]: This is a lot to follow. Just wondering, do you have any tables of experimental versus predicted (with this equation) data size for (T)RCFModelSize

File: src/main/java/org/opensearch/ad/MemoryTracker.java (hunk 1 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/MemoryTracker.java:238]: 1040 is the size of BoundingBoxFloat's fields unrelated to tree size (255 nodes in our formula).  I measure it from heap dump.

File: src/main/java/org/opensearch/ad/MemoryTracker.java (hunk 2 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/MemoryTracker.java:239]: It is the ternary operator in java:

booleanExpression ? expression1 : expression2

File: src/main/java/org/opensearch/ad/MemoryTracker.java (hunk 3 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/MemoryTracker.java:242]: yes, https://gist.github.com/kaituo/77b985921eda631a534157e8e11f6c57

Comment (jmazanec15) [reply src/main/java/org/opensearch/ad/MemoryTracker.java:242]: Nice, those look pretty good. How accurate does this estimate need to be? In other words, whats the margin for error?

Comment (kaituo) [reply src/main/java/org/opensearch/ad/MemoryTracker.java:242]: We don't have a formal margin.  My mental margin is 10%.

**Review by sudiptoguha — APPROVED** [review]

**Review by jmazanec15 — APPROVED** [review]

**Pull request closed without merging**